### PR TITLE
CI updates

### PR DESCRIPTION
**Pull request by AnjoMan**
* new versions of python have come out, we should test against them
* pyproject.toml is now a standard, we should use it
* adds a check of `__version__`
* use pre-commit for the linting, and include isort /  mypy / flake there

### CODE REVIEW

**Comment by andace** [comment]
@AnjoMan have you thought of using black and adding it to the pre-commit hooks?

**Review by veronicaguo — APPROVED** [review]
LGTM!

File: .github/workflows/python-package.yaml [hunk 1]
```diff
@@ -12,29 +12,58 @@ on:
...
+    needs: unit-test
     runs-on: ubuntu-latest
     strategy:
       fail-fast: false
       matrix:
-        python-version: [3.6, 3.7, 3.8, 3.9]
+        python-version: ["3.7", "3.8", "3.9", "3.11", pypy3.8, pypy3.9]
+        # skipping 3.10 since the unit-test job runs against it
     steps:
     - uses: actions/checkout@v2
```
Comment (andace) [inline .github/workflows/python-package.yaml:58]: Should we use checkout@v3?

**Review by andace — CHANGES_REQUESTED** [review]

File: .github/workflows/python-package.yaml (hunk 1 above)
Comment (AnjoMan) [reply .github/workflows/python-package.yaml]: Good thinking! fixed

**Comment by AnjoMan** [comment]
> @AnjoMan have you thought of using black and adding it to the pre-commit hooks?

I have added black to the CI. i did add a bunch of disables, because black is not really optimized around matrix readibility and there are a lot of matrixes in the code.

**Review by andace — APPROVED** [review]

**Pull request merged**